### PR TITLE
Added support for arbitrary cycles in the multigrid.

### PR DESCRIPTION
**Pull request by nloppi**

### CODE REVIEW

File: pyfr/integrators/multip.py [hunk 1]
```diff
@@ -14,28 +14,38 @@ def __init__(self, backend, systemcls, rallocs, mesh, initsoln, cfg):
...
-        self.cycles = list(cfg.getliteral(sect, 'cycle'))
-        self.nlvls = len(self.cycles) // 2 + 1
-        self.levels = (list(range(self.nlvls - 1)) +
-                       list(range(self.nlvls - 1, -1, -1)))
-        self.level = 0
-
-        if self.nlvls > order:
-            raise ValueError('The number of multigrid levels cannot exceed '
+        self.cycle = [c[0] for c in list(cfg.getliteral(sect, 'cycle'))]
+        self.csteps = [c[1] for c in list(cfg.getliteral(sect, 'cycle'))]
```
Comment (FreddieWitherden) [inline pyfr/integrators/multip.py]: self.cycle, self.csteps = zip(*cfg.getliteral(sec, 'cycle'))

File: pyfr/integrators/multip.py [hunk 2]
```diff
@@ -14,28 +14,38 @@ def __init__(self, backend, systemcls, rallocs, mesh, initsoln, cfg):
...
+        self.levels = list(set(self.cycle))[::-1]
+        self.nlvls = len(self.levels)
+        self.level = self.order
+
+        if any(l > self.order for l in self.cycle):
```
Comment (FreddieWitherden) [inline pyfr/integrators/multip.py]: if max(self.cycle) > self.order

File: pyfr/integrators/multip.py [hunk 3]
```diff
@@ -14,28 +14,38 @@ def __init__(self, backend, systemcls, rallocs, mesh, initsoln, cfg):
         sect = 'solver-dual-time-integrator-multip'
 
         # Get the solver order
-        order = cfg.getint('solver', 'order')
+        self.order = cfg.getint('solver', 'order')
```
Comment (FreddieWitherden) [inline pyfr/integrators/multip.py]: Probably not the best idea.

File: pyfr/integrators/multip.py [hunk 4]
```diff
@@ -185,21 +195,20 @@ def system(self):
...
+        for l in self._mgidxcurr:
+            self._mgregs[l] = [proxylist(self.backend.matrix_bank(em, i)
+                                         for em in self._mgsystem[l].ele_banks)
+                               for i in self._regidx]
 
     def _init_system(self, systemcls, *args):
-        self._mgsystem = [systemcls(*args, nreg=self.nreg + 3,
-                                    cfg=self.mgcfg[i])
-                          for i in range(self.nlvls)]
+        self._mgsystem = {}
```
Comment (FreddieWitherden) [inline pyfr/integrators/multip.py]: Dictionary comprehension.
```
self._mgsystem = {l: systemcls(*args, nreg=self.nreg + 3,
                               cfg=self.mgcfg[self.order - l])
                  for l in self.levels}
```

File: pyfr/integrators/multip.py [hunk 5]
```diff
@@ -185,21 +195,20 @@ def system(self):
...
+        self._mgidxcurr = dict.fromkeys(self.levels, 0)
 
         # Create a proxylist of matrix-banks for each storage register
-        for l in range(self.nlvls):
-            self._mgregs.append(
-                [proxylist(self.backend.matrix_bank(em, i)
-                           for em in self._mgsystem[l].ele_banks)
-                 for i in self._regidx]
-            )
+        for l in self._mgidxcurr:
```
Comment (FreddieWitherden) [inline pyfr/integrators/multip.py]: Dictionary comprehension may also work well here for mgregs.

File: pyfr/integrators/multip.py [hunk 6]
```diff
@@ -14,28 +14,38 @@ def __init__(self, backend, systemcls, rallocs, mesh, initsoln, cfg):
...
-        self.nlvls = len(self.cycles) // 2 + 1
-        self.levels = (list(range(self.nlvls - 1)) +
-                       list(range(self.nlvls - 1, -1, -1)))
-        self.level = 0
-
-        if self.nlvls > order:
-            raise ValueError('The number of multigrid levels cannot exceed '
+        self.cycle = [c[0] for c in list(cfg.getliteral(sect, 'cycle'))]
+        self.csteps = [c[1] for c in list(cfg.getliteral(sect, 'cycle'))]
+        self.levels = list(set(self.cycle))[::-1]
```
Comment (FreddieWitherden) [inline pyfr/integrators/multip.py]: Want ``sorted(set(self.cycle), reverse=True)``

File: pyfr/integrators/multip.py [hunk 7]
```diff
@@ -14,28 +14,38 @@ def __init__(self, backend, systemcls, rallocs, mesh, initsoln, cfg):
...
+
+        if any(l > self.order for l in self.cycle):
+            raise ValueError('The multigrid level orders cannot exceed '
                              'the solution order')
 
+        if any(abs(i - j) > 1 for i, j in zip(self.cycle, self.cycle[1:])):
+            raise ValueError('The orders of consecutive multigrid levels can '
+                             'only change by one')
+
+        if self.cycle[-1] != self.order:
```
Comment (FreddieWitherden) [inline pyfr/integrators/multip.py]: What about self.cycle[0]?

File: pyfr/integrators/multip.py [hunk 8]
```diff
@@ -14,28 +14,38 @@ def __init__(self, backend, systemcls, rallocs, mesh, initsoln, cfg):
...
+            raise ValueError('The multigrid cycle needs to end with the '
+                             'highest (solution) order ')
+
         # Multigrid pseudo-time steps
         dtau = cfg.getfloat('solver-time-integrator', 'pseudo-dt')
         dtaufact = cfg.getfloat(sect, 'dt-fact', 1.0)
-        self.dtaus = [dtau*dtaufact**i for i in range(self.nlvls)]
+        self.dtaus = {}
```
Comment (FreddieWitherden) [inline pyfr/integrators/multip.py]: Dictionary comprehension.

File: pyfr/integrators/multip.py [hunk 9]
```diff
@@ -14,28 +14,38 @@ def __init__(self, backend, systemcls, rallocs, mesh, initsoln, cfg):
...
+        for l in self.levels:
+            self.dtaus[l] = dtau*dtaufact**(self.order - l)
 
         # Generate multiple cfgs for the multigrid systems
         self.mgcfg = [Inifile(cfg.tostr()) for i in range(self.nlvls)]
         for deg in range(1, self.nlvls + 1):
```
Comment (FreddieWitherden) [inline pyfr/integrators/multip.py]: Use range(self.nlvls) and update body accordingly.

Comment (nloppi) [inline pyfr/integrators/multip.py]: Changed it to a dictionary to be consistent.

File: pyfr/integrators/multip.py [hunk 10]
```diff
@@ -185,21 +192,17 @@ def system(self):
...
 
         # Create a proxylist of matrix-banks for each storage register
-        for l in range(self.nlvls):
-            self._mgregs.append(
-                [proxylist(self.backend.matrix_bank(em, i)
-                           for em in self._mgsystem[l].ele_banks)
-                 for i in self._regidx]
-            )
+        self._mgregs = {l: [proxylist(self.backend.matrix_bank(em, i)
+                                     for em in self._mgsystem[l].ele_banks)
```
Comment (FreddieWitherden) [inline pyfr/integrators/multip.py]: Add a space.

File: pyfr/integrators/multip.py [hunk 11]
```diff
@@ -185,21 +192,17 @@ def system(self):
...
+                            for i in self._regidx] for l in self.levels}
```
Comment (FreddieWitherden) [inline pyfr/integrators/multip.py]: Add a newline.

File: pyfr/integrators/multip.py [hunk 12]
```diff
@@ -185,21 +192,17 @@ def system(self):
...
 
     def _init_system(self, systemcls, *args):
-        self._mgsystem = [systemcls(*args, nreg=self.nreg + 3,
-                                    cfg=self.mgcfg[i])
-                          for i in range(self.nlvls)]
+        self._mgsystem = {l: systemcls(*args, nreg=self.nreg + 3,
+                                       cfg=self.mgcfg[l]) for l in self.levels}
```
Comment (FreddieWitherden) [inline pyfr/integrators/multip.py]: Add a newline.

File: pyfr/integrators/multip.py [hunk 13]
```diff
@@ -14,40 +14,47 @@ def __init__(self, backend, systemcls, rallocs, mesh, initsoln, cfg):
...
-        self.nlvls = len(self.cycles) // 2 + 1
-        self.levels = (list(range(self.nlvls - 1)) +
-                       list(range(self.nlvls - 1, -1, -1)))
-        self.level = 0
-
-        if self.nlvls > order:
-            raise ValueError('The number of multigrid levels cannot exceed '
+        self.cycle, self.csteps = zip(*cfg.getliteral(sect, 'cycle'))
+        self.levels = sorted(set(self.cycle), reverse=True)
+        self.nlvls = len(self.levels)
```
Comment (FreddieWitherden) [inline pyfr/integrators/multip.py]: Do we need this?

File: pyfr/integrators/multip.py [hunk 14]
```diff
@@ -14,40 +14,46 @@ def __init__(self, backend, systemcls, rallocs, mesh, initsoln, cfg):
...
         dtaufact = cfg.getfloat(sect, 'dt-fact', 1.0)
-        self.dtaus = [dtau*dtaufact**i for i in range(self.nlvls)]
+        self.dtaus = {l: dtau*dtaufact**(self._order - l) for l in self.levels}
 
         # Generate multiple cfgs for the multigrid systems
-        self.mgcfg = [Inifile(cfg.tostr()) for i in range(self.nlvls)]
-        for deg in range(1, self.nlvls + 1):
-            self.mgcfg[deg - 1].set('solver', 'order', order - (deg - 1))
+        self.mgcfg = {l: Inifile(cfg.tostr()) for l in self.levels}
+        for l in self.levels:
```
Comment (FreddieWitherden) [inline pyfr/integrators/multip.py]: for l, mgcfg in self.mgcfg.items()

File: pyfr/integrators/multip.py [hunk 15]
```diff
@@ -14,40 +14,46 @@ def __init__(self, backend, systemcls, rallocs, mesh, initsoln, cfg):
...
+                             'highest (solution) order ')
+
         # Multigrid pseudo-time steps
         dtau = cfg.getfloat('solver-time-integrator', 'pseudo-dt')
         dtaufact = cfg.getfloat(sect, 'dt-fact', 1.0)
-        self.dtaus = [dtau*dtaufact**i for i in range(self.nlvls)]
+        self.dtaus = {l: dtau*dtaufact**(self._order - l) for l in self.levels}
 
         # Generate multiple cfgs for the multigrid systems
-        self.mgcfg = [Inifile(cfg.tostr()) for i in range(self.nlvls)]
```
Comment (FreddieWitherden) [inline pyfr/integrators/multip.py]: Rename to mgcfgs (plural).

**Comment by FreddieWitherden** [comment]
Also change commit message to something along the lines of "Add support for arbitrary cycles with polynomial multigrid."

**Pull request merged**